### PR TITLE
First steps on improving the AVR ESIL analyzer

### PR DESCRIPTION
**Pull request by killabytenow**
Here follows a pull request for starting a rewriting and refactorization of the AVR anal plugin. Currently is not complete but it establishes the base that I will follow next days/commits.

Anyway is better than previous implementation and now it is executing some little AVR programs.

If you want to test it, doesnt forget to execute the following statements before:

aeim 0x00010000 0x01ff eeprom
aeim 0x00010200 0xffff sram

### CODE REVIEW

**Comment by Maijin** [comment]
Doesn't build on windows -> https://ci.appveyor.com/project/radare/radare2-shvdd/build/0.10.3.7088#L2980
Doesn't build on Linux/GCC -> https://travis-ci.org/radare/radare2/jobs/160256295#L285

**Comment by radare** [comment]
do not merge EVER a branch you are working on with master, otherwise reviewing gets much more anoying

File: libr/anal/p/anal_avr.c [hunk 1]
```diff
@@ -17,35 +17,277 @@
...
+	int	pc_mask;
+	int	pc_size;
+
+	int	eeprom_size;
+} CPU_MODEL;
+
+typedef void (*inst_handler_t) (RAnal *anal, RAnalOp *op, ut64 addr, const ut8 *buf, int len, int *fail, CPU_MODEL *cpu);
+
+typedef struct _opcodes_tag_ {
+	char		*name;
```
Comment (radare) [inline libr/anal/p/anal_avr.c]: dont indent variable names, and if you do never use tabs.

File: libr/anal/p/anal_avr.c [hunk 2]
```diff
@@ -17,35 +17,277 @@
 
 #define	AVR_SOFTCAST(x,y) (x+(y*0x100))
 
+typedef struct _cpu_models_tag_ {
+	char	*model;
+	int	pc_bits;
+	int	pc_mask;
+	int	pc_size;
+
```
Comment (radare) [inline libr/anal/p/anal_avr.c]: empty line

File: libr/anal/p/anal_avr.c [hunk 3]
```diff
@@ -17,35 +17,277 @@
...
+	int		mask;
+	int		selector;
+	inst_handler_t	handler;
+	int		cycles;
+	int		size;
+	int		type;
+} OPCODE_DESC;
+
+#define CPU_MODEL_DECL(model, pc_bits, eeprom_sz)						\
+					{							\
```
Comment (radare) [inline libr/anal/p/anal_avr.c]: so many indentation here

File: libr/anal/p/anal_avr.c [hunk 4]
```diff
@@ -17,35 +17,277 @@
...
+	CPU_MODEL_DECL ("ATmega48",   11, 512),
+	CPU_MODEL_DECL ("ATmega8",    12, 512),
+	CPU_MODEL_DECL ("ATmega88",   12, 512),
+	CPU_MODEL_DECL ("ATmega168",  13, 512),
+	CPU_MODEL_DECL ("ATmega640",  16, 512),
+	CPU_MODEL_DECL ("ATmega1280", 16, 512),
+	CPU_MODEL_DECL ("ATmega1281", 16, 512),
+	CPU_MODEL_DECL ("ATmega2560", 22, 512),
+	CPU_MODEL_DECL ("ATmega2561", 22, 512),
+	CPU_MODEL_DECL ((char *) 0,   16, 512) };
```
Comment (radare) [inline libr/anal/p/anal_avr.c]: close the brace in the line below

File: libr/anal/p/anal_avr.c [hunk 5]
```diff
@@ -17,35 +17,277 @@
...
+		"$o,vf,=,"			// overflow flag
+		"vf,nf,^,sf,=",			// sign flag : xor(vf, nf)
+		r, d, r, d);
+}
+
+// ALIAS: Opcode (clr) when Rra and Rrb are the same
+INST_HANDLER (eor) {
+	int d = ((buf[0] & 0xf0) >> 4) | ((buf[1] & 1) << 4);
+	int r = (buf[0] & 0xf) | ((buf[1] & 2) << 3);
+
```
Comment (radare) [inline libr/anal/p/anal_avr.c]: empty line

File: libr/anal/p/anal_avr.c [hunk 6]
```diff
@@ -17,35 +17,277 @@
...
+}
+
+INST_HANDLER (nop) {
+}
+
+INST_HANDLER (out) {
+	op->type2 = 1;
+	op->val = (buf[0] & 0x0f) | (((buf[1] >> 1) & 0x03) << 4);
+	
+	int r = ((buf[0] >> 4) & 0x0f) | ((buf[1] & 0x01) << 4);
```
Comment (radare) [inline libr/anal/p/anal_avr.c]: can we have this in the first line? it is good practice to define variables at the begining of the function

File: libr/anal/p/anal_avr.c [hunk 7]
```diff
@@ -17,35 +17,277 @@
...
+
+	// target offset
+	offset = ((((buf[1] & 0xf) << 8) | buf[0]) + 1) << 1;
+	if (offset & 0x100)
+		offset |= 0xfffff000;
+	op->jump = op->addr + offset;
+	op->fail = 0;
+
+	// cycles!
+	if (strncasecmp (anal->cpu, "ATtiny", 6)) {
```
Comment (radare) [inline libr/anal/p/anal_avr.c]: iwould do the "true" block first, so prefix that condition with ! to make clear that

File: libr/anal/p/anal_avr.c [hunk 8]
```diff
@@ -17,35 +17,277 @@
...
+		"=[%d],"		// store ret@ in stack
+		"sp,-%d,+,"		// decrement stack pointer
+		"sp,=,"			// store SP
+		"%d,pc,=",		// jump!
+		cpu->pc_size - 1, cpu->pc_size,
+		cpu->pc_size, op->jump);
+}
+
+INST_HANDLER (ret) {
+	if (cpu->pc_size > 2)	// if we have a bus bigger than 16 bit
```
Comment (radare) [inline libr/anal/p/anal_avr.c]: add braces

File: libr/anal/p/anal_avr.c [hunk 9]
```diff
@@ -17,35 +17,277 @@
...
+	// op->family = R_ANAL_OP_FAMILY_PRIV;
+
+	// RETI: The I-bit is cleared by hardware after an interrupt
+	// has occurred, and is set by the RETI instruction to enable
+	// subsequent interrupts
+	r_strbuf_append (&op->esil, ",1,if,=");
+}
+
+INST_HANDLER (rjmp) {
+	register int32_t offset;
```
Comment (radare) [inline libr/anal/p/anal_avr.c]: dont use compiler hints like register. they are mostly ignored in modern compiler optimizations

Comment (radare) [inline libr/anal/p/anal_avr.c]: also, use ut32 instead of int32_t, the r2 way

File: libr/asm/p/asm_avr.c [hunk 10]
```diff
@@ -412,11 +412,18 @@ RAsmPlugin r_asm_plugin_avr = {
 	.name = "avr",
 	.arch = "avr",
 	.license = "GPL",
-	.bits = 8|16,
+	.bits = 8|16|32|64,
```
Comment (radare) [inline libr/asm/p/asm_avr.c]: there's no 64bit avr, so drop this,. same for 32 and .. maybe 16 too, because the only 16 bit register is the pc and sp iirc

**Comment by radare** [comment]
also. your code doesnt compiles. here's the error according to travis:

screen shot 2016-09-15 at 22 11 48

File: libr/anal/p/anal_avr.c (hunk 4 above)
Comment (killabytenow) [inline libr/anal/p/anal_avr.c]: it was a recommendation of sys/indent ....

File: libr/anal/p/anal_avr.c (hunk 1 above)
Comment (killabytenow) [inline libr/anal/p/anal_avr.c]: Fixed in new commit

File: libr/anal/p/anal_avr.c (hunk 2 above)
Comment (killabytenow) [inline libr/anal/p/anal_avr.c]: removed

File: libr/anal/p/anal_avr.c (hunk 3 above)
Comment (killabytenow) [inline libr/anal/p/anal_avr.c]: removed

File: libr/anal/p/anal_avr.c (hunk 5 above)
Comment (killabytenow) [inline libr/anal/p/anal_avr.c]: fixed

File: libr/anal/p/anal_avr.c (hunk 6 above)
Comment (killabytenow) [inline libr/anal/p/anal_avr.c]: fixed

File: libr/anal/p/anal_avr.c (hunk 7 above)
Comment (killabytenow) [inline libr/anal/p/anal_avr.c]: A ! was missing

File: libr/anal/p/anal_avr.c (hunk 8 above)
Comment (killabytenow) [inline libr/anal/p/anal_avr.c]: added

File: libr/anal/p/anal_avr.c (hunk 9 above)
Comment (killabytenow) [inline libr/anal/p/anal_avr.c]: removed

Comment (killabytenow) [inline libr/anal/p/anal_avr.c]: replaced

File: libr/asm/p/asm_avr.c (hunk 10 above)
Comment (killabytenow) [inline libr/asm/p/asm_avr.c]: removed

**Comment by radare** [comment]
sys/indent.sh is a crappy thing but yeah, sometimes it works :P

**Comment by radare** [comment]
looks like very basic avr tests are broken now

screen shot 2016-09-16 at 10 02 35

**Comment by radare** [comment]
can you squash and rebase your commits on top of master?

**Comment by radare** [comment]
Closing this PR because its deprecated by the new one

**Pull request closed without merging**